bin/subst.ml

```diff
@@ -299,111 +299,142 @@ let make_watermark_map ~commit ~version ~dune_project ~info =
+        Package.Name.Map.find packages name
+      in
+      let metadata_from_dune_project () = Dune_project.info dune_project.project in
+      let metadata_from_matching_package () =
+        match package_named_after_project with
+        | Some pkg -> Ok (Package.info pkg)
+        | None ->
+          Error
+            (User_error.make
+               ~loc
+               [ Pp.textf "Package %s doesn't exist." (Package.Name.to_string name) ])
+      in
+      let version = Dune_project.dune_version dune_project.project in
+      if version >= (3, 0)
+      then metadata_from_dune_project ()
+      else if version >= (2, 8)
+      then (
+        match metadata_from_matching_package () with
+        | Ok p -> p
+        | Error _ -> metadata_from_dune_project ())
+      else User_error.ok_exn (metadata_from_matching_package ())
     in
-    let metadata_from_dune_project () = Dune_project.info dune_project.project in
-    let metadata_from_matching_package () =
-      match package_named_after_project with
-      | Some pkg -> Ok (Package.info pkg)
-      | None ->
-        Error
-          (User_error.make
-             ~loc
-             [ Pp.textf "Package %s doesn't exist." (Package.Name.to_string name) ])
+    let watermarks = make_watermark_map ~commit ~version ~dune_project ~info in
+    Dune_project.subst ~map:watermarks ~version dune_project;
+    let opam_package_files =
+      Dune_project.packages dune_project.project
+      |> Package.Name.Map.fold ~init:Path.Source.Set.empty ~f:(fun package acc ->
+        Path.Source.Set.add acc (Package.opam_file package))
     in
-    let version = Dune_project.dune_version dune_project.project in
-    if version >= (3, 0)
-    then metadata_from_dune_project ()
-    else if version >= (2, 8)
-    then (
-      match metadata_from_matching_package () with
-      | Ok p -> p
-      | Error _ -> metadata_from_dune_project ())
-    else User_error.ok_exn (metadata_from_matching_package ())
-  in
-  let watermarks = make_watermark_map ~commit ~version ~dune_project ~info in
-  Dune_project.subst ~map:watermarks ~version dune_project;
-  let opam_package_files =
-    Dune_project.packages dune_project.project
-    |> Package.Name.Map.fold ~init:Path.Source.Set.empty ~f:(fun package acc ->
-      Path.Source.Set.add acc (Package.opam_file package))
-  in
-  List.iter files ~f:(fun path ->
-    if is_a_source_file path && not (Path.Source.equal path Dune_project.filename)
-    then subst_file path ~map:watermarks opam_package_files)
+    List.iter files ~f:(fun path ->
+      if is_a_source_file path && not (Path.Source.equal path Dune_project.filename)
+      then subst_file path ~map:watermarks opam_package_files))
 ;;
 
 let subst () =
-  match
-    Sys.readdir "." |> Array.to_list |> String.Set.of_list |> Vcs.Kind.of_dir_contents
-  with
-  | None -> Fiber.return ()
-  | Some kind -> Memo.run (subst { kind; root = Path.root })
+  Sys.readdir "."
```

Comment:
why do we need `Sys.readdir "."` at all if we get the source root in the function above?

Would it be better to find the root first and then try to find the VCS kind? what happens if `subst` is not run from the VCS root?

Reply:
Yeah, we can call `Source_tree.nearest_vcs`. Will leave a TODO